### PR TITLE
[GEOS-11621] OGC API Services Not Honoring the Skip Misconfigured Layers Setting

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-11621](https://osgeo-org.atlassian.net/browse/GEOS-11621) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

Got features test working

added maps collections

added coverages

cleanup

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java [hunk 1]
```diff
@@ -67,6 +72,32 @@ public void testCollectionsJson() throws Exception {
...
     }
 
+    @Test
+    public void testSkipMisconfigured() throws Exception {
+        // enable skipping of misconfigured layers
+        GeoServerInfo global = getGeoServer().getGlobal();
+        global.setResourceErrorHandling(ResourceErrorHandling.SKIP_MISCONFIGURED_LAYERS);
+        getGeoServer().save(global);
+        // manually misconfigure one layer
+        FeatureTypeInfo misconfigured =
```
Comment (aaime) [inline src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java:82]: At this point you're making it recognizable, but you're not really misconfiguring it, just making it recognizable... am I misreading this?

File: src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java [hunk 2]
```diff
@@ -67,6 +72,32 @@ public void testCollectionsJson() throws Exception {
...
+        misconfigured.setOverridingServiceSRS(true);
+        misconfigured.getResponseSRS().addAll(Arrays.asList("3857", "32632"));
+        getCatalog().save(misconfigured);
+
+        DocumentContext json = getAsJSONPath("ogc/features/v1/collections", 200);
+        assertEquals(
+                "Misconfigured",
+                ((LinkedHashMap) ((JSONArray) json.read("collections")).get(16)).get("title"));
+
+        misconfigured.setLatLonBoundingBox(null);
```
Comment (aaime) [inline src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java:95]: This is the actual misconfiguration.

File: src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java [hunk 3]
```diff
@@ -67,6 +72,32 @@ public void testCollectionsJson() throws Exception {
...
+        getCatalog().save(misconfigured);
+        DocumentContext json2 = getAsJSONPath("ogc/features/v1/collections", 200);
+        assertNull(((JSONArray) json2.read("collections")).get(16));
```
Comment (aaime) [inline src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java:98]: So what happens here... is the feature type being returned without the full configuration? It should really be skipped, not showing in the list (so I believe the iteration code should do a sort of loop, check the next, is it sane, if not move to the next, until one is sane or the iteration is completed).

Same goes for all services.

**Review by aaime — CHANGES_REQUESTED** [review]

File: src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java (hunk 1 above)
Comment (turingtestfail) [reply src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java:80]: Fixed, removed unnecessary changes

File: src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java (hunk 2 above)
Comment (turingtestfail) [reply src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java:87]: Yes, added clarifying comments

File: src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java (hunk 3 above)
Comment (turingtestfail) [reply src/community/ogcapi/ogcapi-features/src/test/java/org/geoserver/ogcapi/v1/features/CollectionsTest.java]: Changed for all services

File: src/community/ogcapi/ogcapi-coverages/src/main/java/org/geoserver/ogcapi/v1/coverages/CollectionsDocument.java [hunk 4]
```diff
@@ -99,6 +96,35 @@ public CollectionDocument next() {
...
+            CoverageInfo coverage, CloseableIterator<CoverageInfo> coverages) {
+        try {
+            List<String> crs =
+                    CoveragesService.getCoverageCRS(coverage, Collections.singletonList("#/crs"));
+            CollectionDocument collection = new CollectionDocument(geoServer, coverage, crs);
+            return collection;
+        } catch (Exception e) {
+            if (skipInvalid) {
+                LOGGER.log(Level.WARNING, "Skipping coverage type " + coverage.prefixedName());
+                if (coverages.hasNext()) {
```
Comment (aaime) [inline src/community/ogcapi/ogcapi-coverages/src/main/java/org/geoserver/ogcapi/v1/coverages/CollectionsDocument.java:109]: This works if there is one collection failing and the next is working. But it should work even if 9 out of 10 of collections are broken. The code should look as follows (restructuring this method and the one calling it):

```java
while(coverages.hasNext()) {
   try {
        CoverageDocument next = getCollectionDocument();
        ...
   } catch(Exception e) {
         if (skipExceptions) // log and allow next iteration in while loop
         else rethrow ...
   }
}

Comment (aaime) [reply src/community/ogcapi/ogcapi-coverages/src/main/java/org/geoserver/ogcapi/v1/coverages/CollectionsDocument.java:109]: Ah, recursion actually fetches the next one. Works, but it's limited by the max stack depth

Comment (turingtestfail) [reply src/community/ogcapi/ogcapi-coverages/src/main/java/org/geoserver/ogcapi/v1/coverages/CollectionsDocument.java]: Changed to non-recursion version

**Review by aaime — APPROVED** [review]

**Pull request merged**